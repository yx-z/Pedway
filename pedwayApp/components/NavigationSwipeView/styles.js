import {StyleSheet} from 'react-native';

const styles = StyleSheet.create({
  outerContainer: {
    backgroundColor: '#4185f4',
    height: 140,
    zIndex: 10,
  },
  innerContainer: {
    flex: 1,
    justifyContent: 'center',
  },
  activityIndicator: {
  },
  rowFlexContainer: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
  },
  iconOuterContainer: {
    width: 80,
    height: 80,
  },
  iconInnerContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  colFlexContainer: {
    flex: 1,
    flexDirection: 'column',
    alignItems: 'flex-start',

  },
  instructionIcon: {
    fontSize: 35,
    color: 'white',
  },
  instructionLabel: {
    fontSize: 27,
    color: 'white',
    marginRight: 6,
  },
  distanceLabel: {
    fontSize: 13,
    color: 'white',
    marginTop: 10,
    marginBottom: 6,

  },
});

export default styles;
