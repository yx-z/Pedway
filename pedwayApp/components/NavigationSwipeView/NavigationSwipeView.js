import React, {Component} from 'react';
import styles from './styles';
import {TouchableOpacity, View, ActivityIndicator, Text} from 'react-native';
import Icon from 'react-native-vector-icons/FontAwesome';
import Swiper from 'react-native-swiper';

// map of way types to icon names
// for more information checkout https://github.com/GIScience/openrouteservice-docs/blob/master/README.md
let vectorIconDisplayList = [
  'long-arrow-left',
  'long-arrow-right',
  'long-arrow-left',
  'long-arrow-right',
  'long-arrow-left',
  'long-arrow-right',
  'long-arrow-up',
  'long-arrow-up',
  'long-arrow-up',
  'long-arrow-down',
  'flag',
  'home',
  'long-arrow-left',
  'long-arrow-right',
];

let isProgrammaticallyUpdatingIndex = false;
/**
 * Swiper view based component that displays a list of navigation instructions
 * User can swipe to view the instructions
 * If user is centered on himself/herself(in focus mode), while user is walking, this swiper would swipe automatically
 * to display the most up-to-date instruction
 */

export default class NavigationSwipeView extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      navigationData: [],
      currentIndex: 0,
      dataRequested: false,
      previousIndex: 0,
    };
    this.updateState = this.updateState.bind(this);
    this.onIndexChanged = this.onIndexChanged.bind(this);
    this.updateSwiperViewIndex = this.updateSwiperViewIndex.bind(this);
    this.onMomentumScrollEnd = this.onMomentumScrollEnd.bind(this);
  }

  componentDidMount() {
    this.updateState(this.props);
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      // previousIndex: this.state.currentIndex,
    }, () => {
      if (nextProps.navigationData !== this.state.navigationData) {
        this.updateState(nextProps);
      }
    });
  }

  updateState(inputProps) {
    this.updateSwiperViewIndex(0);
    this.setState({
      navigationData: inputProps.navigationData,
      dataRequested: inputProps.navigationDataRequested,
      currentIndex: 0,
    });

    // set the highlight segment for the path in groundmap to the initial segment
    try {
      let route = inputProps.navigationData['data']['routes'][0];
      let wayPoint = route['segments'][0]['steps'][0]['way_points'];
      this.props.updateSegmentStartEndCallback(wayPoint[0], wayPoint[1]);
    } catch (e) {}
  }

  /**
   * Takes in a 0 indexed index variable, and update the swiper view's index to that variable
   * @param idx
   */
  updateSwiperViewIndex(idx) {
    if (this.state.dataRequested) {
      isProgrammaticallyUpdatingIndex = true;
      this.swiper.scrollBy(idx - this.state.currentIndex + this.state.previousIndex, true);
    }
  }

  /**
   * callBack function that is triggered whenever the swiper view's index changes
   * this function also calls the corresponding function to update the highlighted segment on the mapView
   * @param inputIndex
   */
  onIndexChanged(inputIndex) {
    let acutalIndex = inputIndex + this.state.previousIndex;
    try {
      this.setState({
        currentIndex: acutalIndex,
      });

      let route = this.state.navigationData['data']['routes'][0];
      let wayPoint = route['segments'][0]['steps'][acutalIndex]['way_points'];
      this.props.updateSegmentStartEndCallback(wayPoint[0], wayPoint[1]);
    } catch (e) {
    }
  }

  /**
   * listener that is called when the scroll animation ended
   * we use the global isProgrammaticallyUpdatingIndex is check if this is due to a user swipe or not
   * if it is due to a user swipe, to need to set the focus mode of the groundmapView to false
   */
  onMomentumScrollEnd() {
    // if user is scrolling it, we need to unlock the mapView to let it not in focus
    if (!isProgrammaticallyUpdatingIndex) {
      this.props.setMapInFocus(false);
    }
    isProgrammaticallyUpdatingIndex = false;
  }

  render() {
    if (this.state.dataRequested === false) {
      return (
        <View style={styles.outerContainer}>
          <View style={styles.innerContainer}>
            <ActivityIndicator
              size='large'
              color='#fff'
              animating={true}
              style={styles.activityIndicator}
            />
          </View>
        </View>
      );
    } else {
      // right now using the first route
      let navigationData = [];
      try {
        navigationData = this.state.navigationData['data']['routes'][0]['segments'][0]['steps'];
      } catch (e) {
      }
      let swiperViewData = navigationData.map((item, idx)=>{
        return (
          <View style={styles.rowFlexContainer} key={idx}>
            <View style={styles.iconOuterContainer}>
              <View style={styles.iconInnerContainer}>
                <Icon name={vectorIconDisplayList[item['type']]} style={styles.instructionIcon}/>
              </View>
            </View>
            <View style={styles.colFlexContainer}>
              <Text style={styles.instructionLabel} numberOfLines={3}>{item['instruction']}</Text>
              <Text style={styles.distanceLabel}>{ + (idx + 1) + '/' +
              this.state.navigationData['data']['routes'][0]['segments'][0]['steps'].length +
              '  ' + item['distance'] + 'm'}</Text>
            </View>
          </View>
        );
      });
      return (
        <View style={styles.outerContainer}>
          <View style={styles.innerContainer}>
            <Swiper
              loop={false}
              showsButton={true}
              showsPagination={false}
              onIndexChanged={this.onIndexChanged}
              onMomentumScrollEnd={this.onMomentumScrollEnd}
              ref={(swiper) => {
                this.swiper = swiper;
              }}>
              { swiperViewData }
            </Swiper>
          </View>
        </View>
      );
    }
  }
}
