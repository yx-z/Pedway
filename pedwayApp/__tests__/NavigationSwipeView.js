import NavigationSwipeView from '../components/NavigationSwipeView/NavigationSwipeView';
import testRenderer from 'react-test-renderer';
import React from 'react';


test('Check if our NavigationSwipeView renders correctly', () => {
  const treeRendered = testRenderer.create(<NavigationSwipeView />).toJSON();
  expect(treeRendered).toMatchSnapshot();
});

