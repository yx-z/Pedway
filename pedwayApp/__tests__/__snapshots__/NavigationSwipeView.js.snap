// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check if our NavigationSwipeView renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#4185f4",
      "height": 140,
      "zIndex": 10,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "flex": 1,
            "position": "relative",
          },
          undefined,
        ]
      }
    />
  </View>
</View>
`;
